### PR TITLE
Feature/1960 last success and status failed

### PR DESCRIPTION
**Pull request by kzangeli**
Removed `timesFailed` and added `lastSuccess` for subscriptions.

FIXED:
_One thing is missing in the PR: it seems that lastFailure and lastSuccess have become mutually exclusive. Everything still works (as the newer of the two is always present in the database) but this was not my intention and I would like to ask for some help to fix this problem (or we decide that it is OK as is).
Because of this problem, the documentation is unfinished.
To be decided ..._

### CODE REVIEW

File: src/lib/ngsiNotify/QueueNotifier.cpp [hunk 1]
```diff
@@ -62,29 +62,39 @@ int QueueNotifier::start()
 */
 void QueueNotifier::sendNotifyContextRequest
```
Comment (kzangeli) [inline src/lib/ngsiNotify/QueueNotifier.cpp]: Only style-guide fixes in this file

File: src/lib/mongoBackend/mongoSubCache.cpp [hunk 2]
```diff
@@ -480,88 +481,154 @@ void mongoSubCacheRefresh(const std::string& database)
...
+/* ****************************************************************************
+*
+* mongoSubCountersUpdate - update subscription counters and timestamps in mongo
+*
+* FIXME P4:
+*   This function the way it is implemented right now might perform FIVE consecutive
+*   updates in mongo.
+*   There must be some faster way to do this:
+*     1. Read first, update only what's necessary OR
+*     2. Do a 'mega update' with all five fields (have the functions concatenate to final mongo 'query')
```
Comment (fgalan) [inline src/lib/mongoBackend/mongoSubCache.cpp]: I was thinking on this at the very begining, to "package" count and lastNofication updates into just one operation instead 2...

The fact is that given the filtering condition for each update needs to be evaluated "locally" for each lastX field, you need to split in several places. 

Thus, probably this FIXME should be removed.

File: src/lib/ngsiNotify/QueueNotifier.cpp (hunk 1 above)
Comment (fgalan) [inline src/lib/ngsiNotify/QueueNotifier.cpp]: Ack. NTC

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 3]
```diff
@@ -1759,10 +1759,9 @@ A `notification` object contains the following subfields:
 + `timesSent` (not editable, only present in GET operations): Number of notifications sent due to
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in c44e821

File: src/lib/mongoBackend/mongoUpdateSubscription.cpp [hunk 4]
```diff
@@ -488,20 +488,28 @@ static long long setLastFailure(const BSONObj& subOrig, CachedSubscription* subC
 
 /* ****************************************************************************
 *
-* setTimesFailed - 
+* setLastSuccess -
 */
-static void setTimesFailed(const BSONObj& subOrig, CachedSubscription* subCacheP, BSONObjBuilder* b)
+static long long setLastSuccess(const BSONObj& subOrig, CachedSubscription* subCacheP, BSONObjBuilder* b)
```
Comment (kzangeli) [inline src/lib/mongoBackend/mongoUpdateSubscription.cpp]: Actually, lastNotification is added to before sending the notification while the other two (lastSuccess/lastFailure) need to wait until after to know the status of the notification and the resulting values are stored in the sub-cache only, to be added to mongo when a sub cache refresh is performed.  lastNotification works very differently.

I could change setLastNotification to return the value of lastNotification but that value is not used, so I really don't see any meaning with it ...

To be further discussed perhaps, but, in my opinion, NTC

Comment (fgalan) [inline src/lib/mongoBackend/mongoUpdateSubscription.cpp]: Maybe a comment to setLastNotification() preamble should be added, e.g.:

"Note that this function returns void, different from setLastFailure() and setLastSucces() functions, which use long long. This is due to <a fragment based in your excelent explanation above>"

Comment (kzangeli) [inline src/lib/mongoBackend/mongoUpdateSubscription.cpp]: ok, no probs

Comment (kzangeli) [inline src/lib/mongoBackend/mongoUpdateSubscription.cpp]: Fixed in 2cfde59

File: doc/manuals/admin/database_model.md [hunk 5]
```diff
@@ -309,8 +308,8 @@ Fields:
 -   **payload**: optional field to store the payload for notification customization functionality in NGSIv2.
 -   **lastFailure**: the time when last notification failure occurred.
     Not present if the subscription has never failed.
--   **timesFailed**: number of consecutive notification errors for the subscription.
-    Not present if the subscription has never failed.
+-   **lastSuccess**: the time when last successful notification occurred.
+    Not present if the subscription has never provoked a notification.
```
Comment (fgalan) [inline doc/manuals/admin/database_model.md]: Maybe is more precise

"Not present if the subscription has never provoked a sucessful notification"

Comment (fgalan) [inline doc/manuals/admin/database_model.md]: Fixed in a3b71b891be0b333a5bb84b68a6f8151b1459f7e

**Comment by fgalan** [comment]
LGTM

**Pull request merged**